### PR TITLE
Multi-arch support

### PR DESCRIPTION
**Pull request by Tim15**
See #237 and #304

Todo:

- [x] add multi-arch build capabilities to the alpine-builder script
- [x]  add multi-arch support to the configuration in some way
- [x]  write multi-arch build code
- [x]  generate library spec file
- [x]  fix builder tagging + pushing for gliderlabs images 

Last tasks
- Testing?
- Which versions to roll out on?

### CODE REVIEW

**Comment by Tim15** [comment]
## Update:

I've gotten it to be able to build different architecture images from one config.
Example:
```
# versions/library-edge/options
export RELEASE="edge"
export MIRROR="http://dl-cdn.alpinelinux.org/alpine"
export PACKAGES="alpine-baselayout,apk-tools,alpine-keys,libc-utils"
export ARCHS=(x86_64 armhf)
export BUILD_OPTIONS=(-b -s -t UTC -r $RELEASE -m $MIRROR -p $PACKAGES)
export TAGS=(alpine:edge)
```
Running `./build versions/library-edge/options` will generate:
```
    ...
    ├── library-3.6
    │   ├── Dockerfile
    │   └── options
    └── library-edge
        ├── Dockerfile
        ├── options
        ├── x86_64
        │   ├── Dockerfile
        │   └── rootfs.tar.xz
        └── armhf
            ├── Dockerfile
            └── rootfs.tar.xz
```
I've updated the appropriate docs, and I'd really appreciate feedback on this setup.
There are a few TODOs, listed above.
Currently, all the older images should build perfectly fine.

WDYT @andyshinn @tianon

**Comment by Tim15** [comment]
### Almost There!

I've got the library spec generator working with multi-arch.
If you run `./build library`, you get:
```
# maintainer: Glider Labs <team@gliderlabs.com> (@gliderlabs)
# autogenerated by https://github.com/gliderlabs/docker-alpine/blob/master/build

GitRepo: https://github.com/gliderlabs/docker-alpine.git

Tags:alpine:3.1
Architectures: amd64
GitCommit: 42d89fb9844b17e50875b97c631dd1ca1591107c
amd64-Directory: versions/library-3.1/x86_64

Tags:alpine:3.2
Architectures: amd64
GitCommit: f78fe34e562a3cd50501a44d41a4384a90d9dd42
amd64-Directory: versions/library-3.2/x86_64

Tags:alpine:3.3
Architectures: amd64
GitCommit: 003bfad270dd87ce9cba3b2782a39229c61932f5
amd64-Directory: versions/library-3.3/x86_64

Tags:alpine:3.4
Architectures: amd64
GitCommit: 9fd08ce7d9040ab265adca36633f09570176a4fa
amd64-Directory: versions/library-3.4/x86_64

Tags:alpine:3.5
Architectures: amd64
GitCommit: 731d755ebe08c13be6df34d2ea2daaacbfc974b0
amd64-Directory: versions/library-3.5/x86_64

Tags:alpine:3.6, alpine:latest
Architectures: amd64
GitCommit: 1df7cd436b81e0c8f6f84beb3bf8031de21829d4
amd64-Directory: versions/library-3.6/x86_64

Tags:alpine:edge
Architectures: amd64, arm32v7
GitCommit: c0f432b0e678c1096e041e7e63a089444ad544c9
amd64-Directory: versions/library-edge/x86_64
arm32v7-Directory: versions/library-edge/armhf
```
Then you just copy that into the `library/alpine` spec file in the `docker-library/official-images` repository to release a new version.

The last step is to figure out what to do with multi-arch images for gliderlabs. I'll reach out to them. They may want to just keep their images as x86_64.

Once that's figured out and I fix a few more things, this PR will be all set.

Apologies to all for clogging up this PR and the related issues with updates, etc. I am using this as a sort of documentation in place. Eventually, if requested, I could put it all in a markdown file in `docs`.

After this PR is good, the maintainers will decide which versions to roll out multi-arch on. Currently the only config I've changed is `library/edge`, which now builds x86_64 and armhf (as you can see above). I think this will require coordination with the maintainers of [`docker-library/official-images`](https://github.com/docker-library/official-images), as they are very careful about upgrades and don't want to break anything.

One thing I wonder is if explicitly specifying only amd64 instead of the default inferred architectures will have implications on the [i386 alpine images](https://hub.docker.com/r/i386/alpine/). I hope it won't break anything.

**Comment by andyshinn** [comment]
Also, does this close #252? It looks like this supersedes #252 and maybe helps #190 as well.

File: build [hunk 1]
```diff
@@ -14,21 +14,40 @@ build() {
...
-				} || true
-			fi
+		[[ "$ARCHS" ]] || { # compatability with older configs that were only written for x86_64 architectures
+				ARCHS=("x86_64")
+		}
+
+		for arch in "${ARCHS[@]}"; do
+			# Generate rootfs
+			mkdir -p "$version_dir/$arch"
+			docker run -e "TRACE=$TRACE" --rm "$BUILDER_IMAGE" -a $arch "${BUILD_OPTIONS[@]}" \
```
Comment (andyshinn) [inline build]: `$arch` should probably be double quoted here.

File: build [hunk 2]
```diff
@@ -85,18 +104,57 @@ push() {
 }
 
 library() {
+
+	convert() {
+		# takes a space-separated list of alpine-linux architectures, and returns a space-separated list of docker architectures
+		local i=0
+		for arch in $@; do
```
Comment (andyshinn) [inline build]: I'm guessing `$@` here should be double quoted as well.

File: build [hunk 3]
```diff
@@ -85,18 +104,57 @@ push() {
...
 		# shellcheck source=versions/library-3.2/options
 		source "$file"
 		local refs
 		refs="$(git ls-remote --exit-code --heads origin rootfs/library-${RELEASE#v})"
 		: "${refs:?}"
-		for tag in "${TAGS[@]}"; do
-			echo "${tag#*:}:" \
-				"git://github.com/gliderlabs/docker-alpine@${refs:0:40}" \
-				"versions/library-${RELEASE#v}"
+		tags=${TAGS[@]}
```
Comment (andyshinn) [inline build]: Instead of assigning a string to an array, you can just expand the array with `tags="${TAGS[*]}"`.

File: build [hunk 4]
```diff
@@ -85,18 +104,57 @@ push() {
...
+		[[ "$ARCHS" ]] || { # compatability with older configs that were only written for x86_64 architectures
+				ARCHS=("x86_64")
+		}
+		a=$(convert ${ARCHS[@]})
```
Comment (andyshinn) [inline build]: Double quotes around `${ARCHS[@]}`.

File: build [hunk 5]
```diff
@@ -85,18 +104,57 @@ push() {
...
+
+		printf "\nTags:" && echo ${tags// /, }
```
Comment (andyshinn) [inline build]: Double quotes around the following 3 echoed variables as well.

**Comment by Tim15** [comment]
Fixed the linting issues, added `shopt -s globstar`. I'll need to test the commit function tomorrow. Hopefully that fixes it.

**Comment by Tim15** [comment]
I think this supersedes #252, because this is more fleshed out and complete. It should allow us to add aarch64 (#190) or any other [currently supported architecture](https://github.com/Tim15/docker-alpine/blob/51e4e16e6f0d0846170924eaa91f0d00d67afb69/build#L115-L120) in both docker and alpine.

**Comment by andyshinn** [comment]
Hmm, it looks like the `commit` function will need some more cleanup than I thought. I'm still thinking about how to fix it up (or rewrite it to loop over the `options` files like the other functions do).

**Comment by Tim15** [comment]
I think the best option is to loop over the `options` files, and then instead of just adding and commiting the one `rootfs.tar.xz`, we add all the arch directories, and then commit them.

**Comment by Tim15** [comment]
That should fix the `commit` function, and still be compatible with old the old build structure.

**Comment by andyshinn** [comment]
This seems reasonable. I'll check it out and test tomorrow.

File: build [hunk 6]
```diff
@@ -85,18 +117,59 @@ push() {
...
 library() {
+
+	convert() {
+		# takes a space-separated list of alpine-linux architectures, and returns a space-separated list of docker architectures
+		local i=0
+		for arch in "$@"; do
+			case "$arch" in # converts Alpine Linux arch strings to Docker arch strings
+				x86_64) echo -n "amd64";;
+				x86) echo -n "i386";;
+				armhf) echo -n "arm32v7";; # TODO: What about arm32v6?
```
Comment (tianon) [inline build]: In Alpine, `armhf` is `arm32v6`, not v7 -- Alpine doesn't officially support v7 directly (only indirectly via v6, which is fine).

File: build [hunk 7]
```diff
@@ -85,18 +117,59 @@ push() {
...
+			esac
+			if [[ i -ne $#-1 ]]; then
+				echo -n " "
+			fi
+			let i=$i+1
+		done
+	}
+
+	echo
+	echo "# maintainer: Glider Labs <team@gliderlabs.com> (@gliderlabs)"
```
Comment (tianon) [inline build]: To convert to the new file format, this line needs to be part of the initial "global" block, ala:
```bash
Maintainers: Glider Labs <team@gliderlabs.com> (@gliderlabs)
GitRepo: https://github.com/gliderlabs/docker-alpine.git
```

File: build [hunk 8]
```diff
@@ -85,18 +117,59 @@ push() {
...
+		tags="${TAGS[*]}"
+		[[ "${ARCHS[*]}" ]] || { # compatability with older configs that were only written for x86_64 architectures
+				ARCHS=("x86_64")
+		}
+		a="$(convert "${ARCHS[@]}")"
+
+		echo
+		echo "Tags: ${tags// /, }"
+		echo "Architectures: ${a// /, }"
+		echo "GitCommit: ${refs:0:40}"
```
Comment (tianon) [inline build]: If these commits are not going to be on the `master` branch, then each block will also need a `GitFetch` line:
```bash
Tags: 3.6, latest
Architectures: foo, bar, baz
GitFetch: refs/heads/rootfs/library-3.6
GitCommit: c6cb2bbffbc92d4c35cdaf584dd1c43a860104ea
Directory: versions/library-3.6
```

**Comment by Tim15** [comment]
Thanks for the review. All fixed now.

**Comment by andyshinn** [comment]
Looks good. But CI is failing to build the armhf variant. Maybe this is just an older Docker issue (I think CircleCI is running 1.9)? I'll research this more. The same builds fine locally.

**Comment by Tim15** [comment]
Yeah, it also fails to build on [play-with-docker.com](play-with-docker.com), which is basically an alpine dev/demo environment. I've done some digging, and it seems that the `alpine-baselayout` package, and the `busybox` package have some install scripts that fail. I haven't looked at busybox yet. However, I've download, extracted, and compared the `.apk` files for both the `x86_64` and the `armhf` variants, and the scripts are exactly the same. The main difference is the shared object libraries they require. (`libc.musl-x86_64.so.1`, and `libc.musl-armhf.so.1` respectively). I wonder if @ncopa has any idea why the scripts fail. 

What's really confusing is that they don't fail locally. I'll keep digging and update you guys on this.

**Comment by andyshinn** [comment]
OK, after some more research it seems this was only really working because of https://docs.docker.com/docker-for-mac/multi-arch/. The trigger scripts for arm fail in normal Linux Docker environments.

**Comment by Tim15** [comment]
I see. The unlucky issue of us both using Macs.

**Comment by Tim15** [comment]
I'll take a look at the `busybox` package failures and see if I see the pattern of the platform-dependent `.so` libs.

**Comment by andyshinn** [comment]
I think worst case is maybe I will just remove `armhf` from the `ARCHS` array for now. The rest of this still feels like the right direction and we can probably figure out a workaround for the triggers.

**Comment by Tim15** [comment]
Yup, `busybox` armhf and x86 variants have the same scripts, different `.PKGINFO` lib dependencies, also in this case a different binary.

@andyshinn Probably a good idea while we figure out this issue.

**Comment by Tim15** [comment]
I think the issue is that `apk` doesn't look for the right musl libc. Instead of looking for the `armhf` variant, it looks in the host's (`alpine-builder`) `/lib` directory, which causes errors with the packages that depend on a arch-specific musl libc (`busybox` and `alpine-baselayout`)

I'll look more later, but here are the filesystem logs for the rootfs directory.
`inotifywait -mr $rootfs` for `mkimg-alpine` script
x86_64: https://paste.ee/p/vf7zD
armhf: https://paste.ee/p/EwO1M

**Comment by Tim15** [comment]
It seems that the `edge` release does not provide `armhf` variants of the musl binaries. When downloaded and extracted, the `edge/armhf` package actually contains the `x86_64` binaries. However, the I know the `3.6` version and I'm assuming the older versions as well have the "real" `armhf` binaries. 

I'll dig some more and try to get a fix tomorrow.

**Comment by Tim15** [comment]
The install scripts from the packages just fail. I've opened an upstream issue: https://bugs.alpinelinux.org/issues/7767

There are alternative methods to still implement multi-arch functionality, such as just automatically downloading the `tar.gz` releases from the server (e.g. `dl-cdn.alpinelinux.org/alpine/$version/releases/$arch/alpine-minirootfs-$version-$arch.tar.gz`) and committing them to git as the rootfs.
See release types [here](https://alpinelinux.org/downloads/). We would probably use the `mini-rootfs` variant.

This method would limit customization, as we would have to extract, change stuff, then re-compress. Maybe that's a good thing for the official images, but would be harder for the gliderlabs variants.

WDYT @andyshinn about alternative methods

**Comment by andyshinn** [comment]
There is a `--no-scripts` option to `apk add`. I wonder if we could add this as a new builder option and just ignore the scripts if they are not doing anything needed?

The scripts that are failing (in my build at least) are:

* https://git.alpinelinux.org/cgit/aports/tree/main/alpine-baselayout/alpine-baselayout.pre-install: Maybe we could just emulate this?
* https://git.alpinelinux.org/cgit/aports/tree/main/alpine-baselayout/alpine-baselayout.post-install: Looks like this does a bunch of init / OpenRC stuff. Also some `chown`ing at the end we could emulate. But I wonder if the OpenRC stuff is even needed since generally containers are not used as an init system.
* https://git.alpinelinux.org/cgit/aports/tree/main/busybox/busybox.trigger: Not quite sure what all this is for.
* https://git.alpinelinux.org/cgit/aports/tree/main/busybox/busybox.post-install: This seems to be part of the trigger and just has a comment that it is needed earlier. But what for?

Might be worth trying.

If we were going to go the route of just storing the upstream tarballs I'd think about just rewriting the script to pull those down and commit them. I'm not sure if extracting, testing, and modifying them makes sense in that case. I'd assume that whatever testing of how they were built is the responsibility of upstream.

**Comment by Tim15** [comment]
I suspect we'll get an unusable system, because `busybox --install` symlinks from each command to the busybox multi-call binary. (e.g. `/bin/ls --> /bin/busybox ls`)

**Comment by Tim15** [comment]
Well, just tried it out, and you can use `pwd`, `cd`, `echo`, `printf`, and shell commands, but commands like `ls` and `ps` give `/bin/sh: ls: not found`. I guess we'll just have to emulate the important stuff and/or wait till they fix `apk`.

**Comment by Tim15** [comment]
> If we were going to go the route of just storing the upstream tarballs I'd think about just rewriting the script to pull those down and commit them. I'm not sure if extracting, testing, and modifying them makes sense in that case.

Just pulling the tarballs and committing them without any changes would limit some of the customization options that we currently have. 

This would probably eliminate:
- `$OMIT_COMMUNITY`
- `$REPO_EXTRA`
- `$ADD_BASELAYOUT`
- `$TIMEZONE`
- `$DISABLE_ROOT_PASSWD`
- `$ADD_APK_SCRIPT`

However, this might be a good thing if the goal is to closely mirror upstream without many changes.

Of course, if these options are very important, we could just extract to a temporary working directory, make the changes as before, and then re-compress to the rootfs.

I will start working on a version that just pulls down the tarballs.

**Pull request merged**